src/coins.cpp

```diff
@@ -1027,27 +1099,166 @@ bool CCoinsViewCache::UpdateScInfo(const CScCertificate& cert, CTxUndo& certUndo
+        assert(cert.quality != scIt->second.scInfo.lastCertificateQuality);
 
-    certUndoEntry.replacedLastCertEpoch = scIt->second.scInfo.lastEpochReferencedByCertificate;
-    certUndoEntry.replacedLastCertHash = scIt->second.scInfo.lastCertificateHash;
+        if (cert.quality > scIt->second.scInfo.lastCertificateQuality)
+        {
+            // we found a better cert
+            //---
+            // remove from balance the former top quality amount
+            CAmount bestCertificateAmount = GetValueOfBackwardTransfers(scIt->second.scInfo.lastCertificateHash);
+            scIt->second.scInfo.balance += bestCertificateAmount;
+            LogPrint("cert", "%s():%d - amount restored into scView (amount=%s, resulting bal=%s) %s\n", __func__, __LINE__,
+                FormatMoney(bestCertificateAmount), FormatMoney(scIt->second.scInfo.balance), scId.ToString());
+
+            if (scIt->second.scInfo.balance < totalAmount)
+            {
+                LogPrint("cert", "%s():%d - Can not update balance %s with amount[%s] for scId=%s, would be negative\n",
+                    __func__, __LINE__, FormatMoney(scIt->second.scInfo.balance), FormatMoney(totalAmount), scId.ToString() );
+                return false;
+            }
 
-    scIt->second.scInfo.balance -= totalAmount;
-    scIt->second.scInfo.lastEpochReferencedByCertificate = cert.epochNumber;
-    scIt->second.scInfo.lastCertificateHash = certHash;
-    scIt->second.flag = CSidechainsCacheEntry::Flags::DIRTY;
+            // previous certificate will not have bacwardtransfers anymore
+            NullifyBackwardTransfers(scIt->second.scInfo.lastCertificateHash, certUndoEntry);
+            
+            // update top-quality certificate data
+            scIt->second.scInfo.lastCertificateHash    = certHash;
+            scIt->second.scInfo.lastCertificateQuality = cert.quality;
+            LogPrint("cert", "%s():%d - cert quality updated in scView (best cert=%s, best q=%d)\n", __func__, __LINE__,
+                scIt->second.scInfo.lastCertificateHash.ToString(), scIt->second.scInfo.lastCertificateQuality);
 
-    LogPrint("cert", "%s():%d - amount removed from scView (amount=%s, resulting bal=%s) %s\n",
-        __func__, __LINE__, FormatMoney(totalAmount), FormatMoney(scIt->second.scInfo.balance), scId.ToString());
+            scIt->second.scInfo.balance -= totalAmount;
+            LogPrint("cert", "%s():%d - amount removed from scView (amount=%s, resulting bal=%s) %s\n",
+              __func__, __LINE__, FormatMoney(totalAmount), FormatMoney(scIt->second.scInfo.balance), scId.ToString());
 
+        }
+        else
+        {
+            // should never happen if certs are ordered by quality in a block
+            LogPrint("cert", "%s():%d - cert quality %d not greater than last seen %d", 
+                __func__, __LINE__, cert.quality, scIt->second.scInfo.lastCertificateQuality);
+            return false;
+        }
+    }
```

Comment:
As discussed offline, this case is temporary, before enforcing certificates ordering in blocks by quality both in consensus and in block formation.
However I believe hard coding such ordering in UpdateScInfo would make in unusable in mempool. I believe this is very bad, since in next future we would want to extend UpdateScInfo checks on CTxMempool::check()